design/StableRegions-20230209.md

```diff
@@ -0,0 +1,224 @@
+ - The `region` table (fixed size, about 262kb).
+
+### representation of values of type `Region`
+
+ - A singleton, heap-allocated object with mutable fields.
+ - While being heap-allocated, the object is also `stable` (can be stored in a `stable var`, etc).
+ - `RegionObject { size_in_pages: Nat64; id: Nat16; vec_capacity: Nat16; vec_ptr: Nat32 }`
+ - Field `size_in_pages` gives the number of pages allocated to the Region.
+ - Field `id` gives the Region's numerical id as an index into the `region` table.
+ - Fields `vec_capacity` and `vec_ptr` work with `size_in_pages`
+   to represent a growable vector that we call the region's **"access
+   vector"** (because "blocks vector" sounds a bit strange, and its
+   used to support O(1) access operations):
+   - the access vector's address is held in `vec_ptr` and it has `vec_capacity` slots.
+   - the first `size_in_pages / 128` slots of `vec_ptr` contain a valid page block ID for the region.
+   - the access vector doubles when it grows.
+   - no region has more than 32k page blocks, so a `Nat16` suffices for `capacity`,
+   - during an upgrade, the access vectors get serialized and deserialized as data `Blobs` (as if no pointers are inside each).
+
+
+### region-blocks relation
+
+The `region-blocks` relation is not materialized into a table in stable memory (doing so with a pre-allocated table would be prohibitively large).
+
+Instead, this relation is represented in two ways at the same time:
+ 1. by the set of heap-allocated region objects, and their access vectors.  The access vectors provide O(1) store and load support.
+ 2. by the `block-region` table and the `region` table, which together are sufficient to recreate all of the heap-allocated region objects.
+
+In ordinary operation, the second feature is not required.  In the event of an upgrade failure, however, it could be vital (See `rebuild`).
+
+### block-region table
+
+ - purpose:
+   - relate a block ID ("page block ID") to its region (if any) and its position in that region (see `rebuild`).
+
+   - NB: The organization of this table is not useful for efficient
+     access calculations of load or store (they require a linear
+     search that would be prohibitively slow).  OTOH, this
+     table is suitable to do a "batch" rebuild of the dynamic heap-allocated vectors
+     in that table, if need be (see `rebuild`).
+
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BlockRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding block ID.
+
+### region table
+
+ - purpose:
+   - relate a region ID to its size in pages.
+
+   - NB: This table exists as a "stable backup" of the per-region fields
+     that exist in the set of region objects, except for the access vectors' elements (see `block-region` table for their data).
+
+ - 32768 entries (statically sized).
+ - 8 bytes per entry.
+ - entry type = `RegionBlocks { size_in_pages: Nat64; }`
```

Comment:
```suggestion
 - entry type = `RegionBlocks { size_in_regions: Nat16; remainder_in_pages: nat8 }`
```
I wonder if you could compress this to 3 bytes per entry: size in (full) regions (at most 32768) + remaining pages (at most 127). But don't trust my arithmetic.
That would save 5 bytes per entry, I think. Or just use two Nat16s and save 4 bytes per entry.

Reply:
I think you might need one more field indicating whether the region is free or allocated. Or some other state to indicate how many regions are in use - the block-region table probably isn't enough since a region may be allocated but have zero blocks (initially). Perhaps @luc-blaeser has a clever idea on how to keep these in a free list when you need to allocate a new region, possible re-using an old one in future. A simple ID counter would do initially but maybe not once GC is added.

Reply:
Sorry, I hadn't seen the comments by you and Luc earlier on. Might be worth writing it down in the design doc.

Reply:
Okay, I'll add a section where I gather the GC-related stuff all together and go through the story we have now.